### DISCUSSION TITLE
A race condition in networking setup

### DISCUSSION THREAD

**Discussion by farcaller** [cloud-hypervisor/cloud-hypervisor#5031] [Q&A] [answered]
**Describe the bug**
Unless I'm misunderstanding something in the code, the tap interface is created on boot()/reboot() api calls. This means the VM starts booting up while the host might not yet had time to finish setting up the network (firewall rules, adding the tap device to the bridge, etc.)

While in the overwhelming majority of the cases it's good enough (because it takes a while for the linux kernel get to the point where it expects network), there are some edge cases where this is troublesome.

One easy option to work this around would be allowing to boot a VM suspended by extending the vm.boot api with an optional suspend flag. This way it should behave very similar to the gdb-based stop-on-boot integration by setting up the devices but not executing anything until the host explicitly resumes it.

**To Reproduce**
1. Set up e.g. a linux bridge `vmbr0` on the host, that allows NAT
2. Boot up the VM
3. Be really slow about plugging the vm's tap0 into vmbr0.
4. Observe that during this time the VM considers the network interface up despite the host side not being functional yet.
 
**Version**

Output of `cloud-hypervisor --version`: v28.0.0

Did you build from source, if so build command line (e.g. features): no

**Comment by rbradford** [comment, marked as answer]
You can create the tap interface in advance, attach it to the bridge, setup iptables etc and then start the VM. e.g. `sudo ip tuntap add mytap0 mode tap` and the specify `--net tap=mytap0` and Cloud hypervisor will use the created TAP device in lieu of creating a new one.
